lib/spack/spack/concretize.py

```diff
@@ -347,26 +347,46 @@ def concretize_compiler_flags(self, spec):
             # running.
             return True
 
-        compiler_match = lambda other: (
-            spec.compiler == other.compiler and
-            spec.architecture == other.architecture)
+        def compiler_match(other, strict=True):
+            if strict:
+                # Other compiler already matches this compiler.
+                return (spec.compiler == other.compiler and
+                        spec.architecture == other.architecture)
+            else:
+                # Other compiler may match this compiler.
+                if other.compiler and other.compiler.concrete:
+                    return (spec.compiler == other.compiler and
+                            spec.architecture == other.architecture)
+                return True
 
         ret = False
         for flag in spack.spec.FlagMap.valid_compiler_flags():
             if flag not in spec.compiler_flags:
-                spec.compiler_flags[flag] = list()
-            try:
-                nearest = next(p for p in spec.traverse(direction='parents')
-                               if (compiler_match(p) and
-                                   (p is not spec) and
-                                   flag in p.compiler_flags))
-                nearest_flags = set(nearest.compiler_flags.get(flag, []))
-                flags = set(spec.compiler_flags.get(flag, []))
-                if (nearest_flags - flags):
-                    spec.compiler_flags[flag] = list(nearest_flags | flags)
-                    ret = True
-            except StopIteration:
-                pass
+                try:
+                    n = next(p for p in spec.traverse(direction='parents')
+                             if (compiler_match(p) and
+                                 (p is not spec) and
+                                 flag in p.compiler_flags))
+
+                    nearest_flags = set(n.compiler_flags[flag])
+                    flags = set()
+
+                    if (nearest_flags - flags) or n.compiler_flags[flag] == []:
+                        spec.compiler_flags[flag] = list(nearest_flags | flags)
+                        ret = True
+
+                except StopIteration:
+                    try:
+                        n = next(p for p in spec.traverse(direction='parents')
+                                 if (compiler_match(p, False) and
+                                     p is not spec))
+                        # Wait for things higher up DAG to settle down.
+                        # Don't incorporate compiler config information until
+                        # DAG information is incorporated, so return here.
+                        return True
+                    except StopIteration:
+                        spec.compiler_flags[flag] = []
+                        ret = True
 
         # Include the compiler flag defaults from the config files
         # This ensures that spack will detect conflicts that stem from a change
```

Comment:
Since this logic hasn't changed, is it possible that compiler flags will be added from the default compiler spec to the already-installed dependency?

Reply:
Bailing out at the beginning of the function if `spec._concrete` resolves my original concern here. But I wanted to make sure: it was your intent that the compiler definition be able to merge into compiler flags explicitly set by the user (although above in this PR you don't want to merge flags from other compiler specs)?